# Milestone 2: Insight Of Data And Feasibility Check Of Our Idea

<hr style="clear:both">

This notebook traces our journey through the feasibility check of our idea : Checking trends in music composition for the movie industry.

**Project Mentor:** [Aoxiang Fan](https://people.epfl.ch/aoxiang.fan) ([Email](mailto:aoxiang.fan@epfl.ch))
**Authors:** [Luca Carroz](https://people.epfl.ch/emilie.carroz), [David Schroeter](https://people.epfl.ch/david.schroeter), [Xavier Ogay](https://people.epfl.ch/xavier.ogay), [Joris Monnet](https://people.epfl.ch/joris.monnet), [Paulo Ribeiro de Carvalho](https://people.epfl.ch/paulo.ribeirodecarvalho)

<hr style="clear:both">

## Storyline

**Title : Stanislas' music dream : Road to Hollywood !**

A 20-year-old aspiring musician, Stanislas, fueled by a passion for the film industry, embarks on a quest to launch his career. His ultimate dream? To hear one of his productions featured in a Hollywood film and become one of the planet's top composers. To increase his chances, he turns to a team of Data Scientists known as LSD.

The "LearningtheSecretsofData" team's mission is to identify trends shared among successful music composers and compositions, ultimately optimizing choices for our young musician. This is not an easy task but the team is driven by the wish of helping Stanislas. How could they provoke a cascADA of successful choices in Stany career.

Which music genre Stany should he focus on? Will this new direction be enough for him to conquers the show business? Maybe he may invest in a ludicrous website to promote himself? Or should he even consider changing Nationality to achieve his goal? Let’s see what’s the plan LSD had concocted for Stanislas.


## Needed data

To help Stanislas with his dream, we will need some specific information about each analysed movie. To support our final answer concerning potential trends in composer for movie industry, we will analyse few questions :

1) Which are the most frequent music genre appearing in movies ?
2) What is the average composer's age at their :
   - first movie appearance ?
   - biggest box office revenue ?
3) How the top composers' career progress over the years ?
4) Where do composers come from ?
5) Does composer's gender matter ?
6) Does having a personal website correlate with the composers' success ?
7) Is there a correlation between box office revenue and movie's playlist popularity ?

List of needed information about movie's **composer** :

- Name
- Birthday
- Gender
- Homepage
- Place of birth
- First appearance in movie credits

List of needed information about composers' **musics** :

- Genre
- Spotify's popularity


## Import

In [1]:
# Import all needed libraries
from helpers import *

# Load autoreload extension
%load_ext autoreload

# Set autoreload mode
%autoreload 2

## Load Data

Only `movie.metadata.tsv` is loaded, since others data sets are not relevant for our data analysis.

Note that no header are given in the raw data sets. We then looked at the documentation and set these headers with meaningful names. Please, find the documentation of data set [here](http://www.cs.cmu.edu/~ark/personas/).

In [2]:
# Please be sure your data is store in the same path
data_path = 'dataset/MovieSummaries/movie.metadata.tsv'

# Load the text file into Pandas DataFrame
movie = load_movies(data_path)

# Display the dataframe
display(movie)

wiki_movieID freebase_movieID  \
0            975900        /m/03vyhn   
1           3196793        /m/08yl5d   
2          28463795       /m/0crgdbh   
3           9363483       /m/0285_cd   
4            261236        /m/01mrr1   
...             ...              ...   
81736      35228177       /m/0j7hxnt   
81737      34980460       /m/0g4pl34   
81738       9971909       /m/02pygw1   
81739        913762        /m/03pcrp   
81740      12476867       /m/02w7zz8   

                                                    name release_date  \
0                                         Ghosts of Mars   2001-08-24   
1      Getting Away with Murder: The JonBenét Ramsey ...   2000-02-16   
2                                            Brun bitter         1988   
3                                       White Of The Eye         1987   
4                                      A Woman in Flames         1983   
...                                                  ...          ...   
81736                           Mermaids: The Body Found   2011-03-19   
81737                                            Knuckle   2011-01-21   
81738                                  Another Nice Mess   1972-09-22   
81739  The Super Dimension Fortress Macross II: Lover...   1992-05-21   
81740                                            Spliced         2002   

       box_office_revenue  runtime                           languages  \
0                14010832     98.0  {"/m/02h40lc": "English Language"}   
1                    <NA>     95.0  {"/m/02h40lc": "English Language"}   
2                    <NA>     83.0  {"/m/05f_3": "Norwegian Language"}   
3                    <NA>    110.0  {"/m/02h40lc": "English Language"}   
4                    <NA>    106.0   {"/m/04306rv": "German Language"}   
...                   ...      ...                                 ...   
81736                <NA>    120.0  {"/m/02h40lc": "English Language"}   
81737                <NA>     96.0  {"/m/02h40lc": "English Language"}   
81738                <NA>     66.0  {"/m/02h40lc": "English Language"}   
81739                <NA>    150.0   {"/m/03_9r": "Japanese Language"}   
81740                <NA>     86.0  {"/m/02h40lc": "English Language"}   

                                               countries  \
0              {"/m/09c7w0": "United States of America"}   
1              {"/m/09c7w0": "United States of America"}   
2                                 {"/m/05b4w": "Norway"}   
3                         {"/m/07ssc": "United Kingdom"}   
4                                {"/m/0345h": "Germany"}   
...                                                  ...   
81736          {"/m/09c7w0": "United States of America"}   
81737  {"/m/03rt9": "Ireland", "/m/07ssc": "United Ki...   
81738          {"/m/09c7w0": "United States of America"}   
81739                              {"/m/03_3d": "Japan"}   
81740                            {"/m/0d060g": "Canada"}   

                                                  genres  
0      {"/m/01jfsb": "Thriller", "/m/06n90": "Science...  
1      {"/m/02n4kr": "Mystery", "/m/03bxz7": "Biograp...  
2      {"/m/0lsxr": "Crime Fiction", "/m/07s9rl0": "D...  
3      {"/m/01jfsb": "Thriller", "/m/0glj9q": "Erotic...  
4                                {"/m/07s9rl0": "Drama"}  
...                                                  ...  
81736                            {"/m/07s9rl0": "Drama"}  
81737  {"/m/03bxz7": "Biographical film", "/m/07s9rl0...  
81738       {"/m/06nbt": "Satire", "/m/01z4y": "Comedy"}  
81739  {"/m/06n90": "Science Fiction", "/m/0gw5n2f": ...  
81740  {"/m/01jfsb": "Thriller", "/m/03npn": "Horror"...  

[81741 rows x 9 columns]

A quick visual insight can be done to get information gain. From the $81,741$ movies, we see that some variables are textual (i.e. `freebase_movieID`, `name`, ...) or numerical (i.e. `box_office_revenue`, `runtime`, ...).

Let's have a more meaningful insight of our current data.

In [3]:
insight(movie)

class  missing_values
wiki_movieID          <class 'int'>            0.00
freebase_movieID      <class 'str'>            0.00
name                  <class 'str'>            0.00
release_date          <class 'str'>            8.44
box_office_revenue    <class 'int'>           89.72
runtime             <class 'float'>           25.02
languages             <class 'str'>            0.00
countries             <class 'str'>            0.00
genres                <class 'str'>            0.00

We see that only few movies ($8,401$) have information about their box office revenue. Also, not all needed attribute are stored in the actual raw data. Since they are important metrics for out analysis, let's try to retrieve as much information as possible via API requests (procedure explained below).

## Clean/Enrich Data

 To enrich our dataset, we use a free to use API ([TMDB](https://www.themoviedb.org/?language=fr)). Some important features are still missing after this procedure. Then, we drop these movies from our analysis.

A specific script has been created to be run once and create our `clean_enrich_movie.csv` dataset. Go to `enrich_data.py` and its linked library `tmdb/tmdb.py` for more explanation on how we retrieved these information. Please note that a personal API key is needed to successfully run the script ([create key](https://developer.themoviedb.org/reference/intro/getting-started)). Make sure to create file `.env` with your API bearer token using the `.env_example` as template.

In [4]:
# TODO update the markdown just above with the final API procedure. What we retrieve and so on.
# Load the cleaned and enriched data set (created via enrich_movie_data.py script)
enhanced_movie = pd.read_pickle('dataset/clean_enrich_movies.pickle')

# Display the loaded dataframe
display(enhanced_movie)

name release_date  \
0                                            Avatar         2009   
1                                           Titanic         1997   
2                                      The Avengers         2012   
3     Harry Potter and the Deathly Hallows – Part 2         2011   
4                    Transformers: Dark of the Moon         2011   
...                                             ...          ...   
8322                              Frankie and Alice         2010   
8323                           Fighting Tommy Riley         2005   
8324                                          Logan         2010   
8325                                  GhettoPhysics         2010   
8326       Hugh Hefner: Playboy, Activist and Rebel         2010   

      box_office_revenue                                   countries  \
0             2782275172  [United States of America, United Kingdom]   
1             2185372302                  [United States of America]   
2             1511757910                  [United States of America]   
3             1328111219  [United States of America, United Kingdom]   
4             1123746996                  [United States of America]   
...                  ...                                         ...   
8322               10670                                    [Canada]   
8323               10514                  [United States of America]   
8324               10474                  [United States of America]   
8325               10200                                          []   
8326               10000                                    [Canada]   

                                                 genres  tmdb_id  \
0     [Thriller, Science Fiction, Adventure, Compute...    19995   
1     [Tragedy, Costume drama, Historical fiction, A...      597   
2                             [Science Fiction, Action]    24428   
3                  [Drama, Mystery, Fantasy, Adventure]    12445   
4      [Alien Film, Science Fiction, Action, Adventure]    38356   
...                                                 ...      ...   
8322                                 [Biography, Drama]    55061   
8323               [LGBT, Sports, Drama, Indie, Boxing]    47534   
8324                       [Family Film, Drama, Comedy]    44010   
8325                    [Political cinema, Documentary]       -1   
8326  [Culture & Society, Biography, Media Studies, ...       -1   

                                              composers  
0     [Composer(id=1729, name='James Horner', birthd...  
1     [Composer(id=1729, name='James Horner', birthd...  
2     [Composer(id=37, name='Alan Silvestri', birthd...  
3     [Composer(id=2949, name='Alexandre Desplat', b...  
4     [Composer(id=18264, name='Steve Jablonsky', bi...  
...                                                 ...  
8322  [Composer(id=70789, name='Andrew Lockington', ...  
8323                                                NaN  
8324                                                NaN  
8325                                                NaN  
8326                                                NaN  

[8327 rows x 7 columns]

Let's have a more deep look at this new dataframe. First, let's have a look of the percentage of missing values now.

In [5]:
insight(enhanced_movie)

class  missing_values
name                 <class 'str'>            0.00
release_date         <class 'str'>            0.00
box_office_revenue   <class 'int'>            0.00
countries           <class 'list'>            0.00
genres              <class 'list'>            0.00
tmdb_id              <class 'int'>            0.00
composers           <class 'list'>           30.92

We also display statistical values to better understand the distribution of few metrics of the data.

In [6]:
# TODO see if we can give better insight of data.
insight_enhance(enhanced_movie)

count              8327.0
mean      48269077.422241
std      112467320.603979
min               10000.0
25%             2105039.0
50%            10815378.0
75%            41050113.0
max          2782275172.0
Name: box_office_revenue, dtype: Float64

There is 30.92% of nan composers

Considering the first composer of the list if multiple have been returned for a movie, we can compute the following statistics on the retrieved data:

	 - There is 0.00% of nan name for composers
	 - There is 18.08% of nan birthday for composers
	 - There is 0.00% of nan gender for composers
	 - There is 72.24% of nan homepage for composers
	 - There is 22.93% of nan place of birth for composers
	 - There is 100.00% of nan first appearance in movie for composers


/Users/pauloribeiro/Desktop/EPFL/Master Data Science/Semestre 1/Applied Data Analysis/ada-2023-project-learningthesecretsofdata/helpers.py:130: FutureWarning: using <function insight_enhance.<locals>.<lambda> at 0x11fe3b0a0> in Series.agg cannot aggregate and has been deprecated. Use Series.transform to keep behavior unchanged.
  composers_no_na_name = composers_no_na.agg(lambda c: c[0].name)
/Users/pauloribeiro/Desktop/EPFL/Master Data Science/Semestre 1/Applied Data Analysis/ada-2023-project-learningthesecretsofdata/helpers.py:131: FutureWarning: using <function insight_enhance.<locals>.<lambda> at 0x10f1fa170> in Series.agg cannot aggregate and has been deprecated. Use Series.transform to keep behavior unchanged.
  composers_no_na_birthday = composers_no_na.agg(lambda c: c[0].birthday)
/Users/pauloribeiro/Desktop/EPFL/Master Data Science/Semestre 1/Applied Data Analysis/ada-2023-project-learningthesecretsofdata/helpers.py:132: FutureWarning: using <function insight_enhance.<locals>.

## Results Expected

Let's try to explain what are the main plots we want to bring to our readers. Do we have in mind also to have a interactive plot, and so on ?

In [7]:
pass

In [8]:
import pandas as pd
test=pd.read_csv('test')
test_without = pd.read_csv('test_without')

nb_test = test['box_office_revenue'].isna().sum()
nb_test_without = test_without['box_office_revenue'].isna().sum()
print(f'Number of movies without box office revenue : {nb_test}')
print(f'Number of movies without box office revenue (without enrich) : {nb_test_without}')

FileNotFoundError: [Errno 2] No such file or directory: 'test'